x/transferinject/ibc_middleware.go

```diff
@@ -0,0 +1,193 @@
+	timeoutHeight clienttypes.Height,
+	timeoutTimestamp uint64,
+	data []byte,
+) (sequence uint64, err error) {
+	packet := new(transfertypes.FungibleTokenPacketData)
+	if err = types.ModuleCdc.UnmarshalJSON(data, packet); err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrUnknownRequest, "unmarshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	if types.MemoAlreadyHasPacketMetadata(packet.Memo) {
+		return 0, errorsmod.Wrapf(errortypes.ErrUnauthorized, "denom metadata already exists in memo")
+	}
+
+	rollapp, err := m.rollappKeeper.ExtractRollappFromChannel(ctx, destinationPort, destinationChannel)
+	if err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrNotFound, "extract rollapp id from packet: %s", err.Error())
+	}
+	// TODO: consider that other chains also might want this feature
+	if rollapp == nil {
+		return m.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+	}
+
+	if hasDenom(rollapp.TokenMetadata, packet.Denom) {
+		return m.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+	}
+
+	denomMetadata, ok := m.bankKeeper.GetDenomMetaData(ctx, packet.Denom)
+	if !ok {
+		return 0, errorsmod.Wrapf(errortypes.ErrNotFound, "denom metadata not found")
+	}
+
+	packet.Memo = types.AddDenomMetadataToMemo(packet.Memo, denomMetadata)
+
+	data, err = types.ModuleCdc.MarshalJSON(packet)
+	if err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrInvalidType, "marshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	return m.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+}
+
+// OnAcknowledgementPacket adds the token metadata to the rollapp if it doesn't exist
+func (m *IBCMiddleware) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	var ack channeltypes.Acknowledgement
+	if err := types.ModuleCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
+		return errorsmod.Wrapf(rtypes.ErrUnknownRequest, "unmarshal ICS-20 transfer packet acknowledgement: %v", err)
+	}
+
+	if !ack.Success() {
+		return m.IBCModule.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+
+	var data transfertypes.FungibleTokenPacketData
+	if err := types.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return errorsmod.Wrapf(errortypes.ErrUnknownRequest, "unmarshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	if len(data.Memo) == 0 {
+		return m.IBCModule.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+
+	memoData, err := types.ParseMemoData(data.Memo)
+	if errors.Is(err, types.ErrMemoUnmarshal) || errors.Is(err, types.ErrMemoDenomMetadataEmpty) {
+		return m.IBCModule.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+	if err != nil {
+		return err
+	}
+
```

Comment:
I don't think there should be an additional if len==0 check
If it's empty, the types.ParseMemoData should return some appropriate error which can then be checked with errors.Is